src/plugins/kibana_utils/common/state_containers/types.ts

```diff
@@ -63,45 +116,92 @@ export interface ReduxLikeStateContainer<
   subscribe: (listener: (state: State) => void) => () => void;
 }
 
+/**
+ * Redux like dispatch
+ * @public
+ */
 export type Dispatch<T> = (action: T) => void;
+/**
+ * Redux like Middleware
+ * @public
+ */
 export type Middleware<State extends BaseState = BaseState> = (
   store: Pick<ReduxLikeStateContainer<State, any>, 'getState' | 'dispatch'>
 ) => (
   next: (action: TransitionDescription) => TransitionDescription | any
 ) => Dispatch<TransitionDescription>;
-
+/**
+ * Redux like Reducer
+ * @public
+ */
 export type Reducer<State extends BaseState> = (
   state: State,
   action: TransitionDescription
 ) => State;
 
+/**
+ * Utility type for inferring state shape from {@link StateContainer}
+ * @public
+ */
 export type UnboxState<
   Container extends StateContainer<any, any>
 > = Container extends StateContainer<infer T, any> ? T : never;
+/**
+ * Utility type for inferring transitions type from {@link StateContainer}
+ * @public
+ */
 export type UnboxTransitions<
   Container extends StateContainer<any, any>
 > = Container extends StateContainer<any, infer T> ? T : never;
 
+/**
+ * @public
+ */
 export type Selector<Result, Args extends any[] = []> = (...args: Args) => Result;
+/**
+ * @public
+ */
 export type PureSelector<State extends BaseState, Result, Args extends any[] = []> = (
   state: State
 ) => Selector<Result, Args>;
+/**
+ * @public
+ */
 export type EnsurePureSelector<T> = Ensure<T, PureSelector<any, any, any>>;
+/**
+ * @public
+ */
```

Comment:
I am pretty happy with generated docs for `state_sync`, but for `state_containers` it is not that great... mostly due how complicate type system of `state_containers` is.. There are a lot of helper types which are exposed publicly and I am not sure if we should try to document them. don't see much usefulness for documenting all of that and is not a quick effort..

@streamich let me know if you have ideas or suggestions.